rfcs: added rfc describing `__consumer_offsets` support (#3767)
redpanda-data/redpanda: docs/rfcs/20220209_consumer_offsets.md

decoding capabilities so that RPK will be able to display human/machine 
readable content of `__consumer_offsets` topic.

### Admin API

There is no need to expose `__consumer_offsets` alias via Admin API. 
We may assume that admin API users should be aware of Redpanda internals and 
we may just operate using `kafka_internal/group` Redpanda specific topic name.

## Corner cases dissected by example.

### Topic `kafka/__consumer_offsets` already exists in the cluster

In this case we will just use whatever is there and ignore an alias.

### Topic `kafka/__consumer_offsets` is being created by user

In this case Redpanda will create underlying `kafka_internal/groups` topic with 
a specified set of options provided by a caller. We may emit WARN log message 
that the topic being created is an internal topic. 

## Detailed design - What needs to change to get there

### Internal topics

We will introduce a function `bool is_internal_topic(const model::topic&)` 
that will return true whenever a topic is internal. This will provide 
compatibility with Kafka internal topics abstraction. Topics are marked as 
internal in  `MetadataResponse`. 

### Topic aliases

> Adding an `std::vector<topic_alias>` to topic table where topic aliases will be 
> registered. We can tolerate the linear lookup const since number of aliases 
> is very limited (currently 2).

Reviewer: ```suggestion
Add an `std::vector<topic_alias>` sequence container to the topic table where topic aliases will be 
registered. We can tolerate the linear lookup const since the number of aliases 
is very limited (currently 2).
```

Reviewer: make sure my "sequence container" edit is correct

```c++
struct  topic_alias {
    model::topic_namespace alias;
    model::topic_namespace source;
};
```

`cluster::metadata_cache` will transparently handle all APIs for aliased topics.

NOTE: `topic_alias` data structure is generic however for simplicity in 
Kafka API handlers we will handle aliases for topics marked as `is_internal` 

### Translation layer

> Adding a data structure that will hold data transformation factor related with given 
> topic alias. Reference to that data structure will be provided to request via 
> `kafka::request_context`. Data transformation is going to be a simple class 
> creating `model::record_batch_reader` instances that will apply simple 
> transformation.

Reviewer: ```suggestion
We must add a data structure that will hold the data transformation factor related to the given 
topic alias. Reference to that data structure will be provided by request via 
`kafka::request_context`. Data transformation will be a simple class, 
creating `model::record_batch_reader` instances that will apply a simple 
transformation.
```

## Drawbacks

The main drawback of the proposed solution is that it must be special cased in 
multiple places in the code. However, creating a generic internal topic alias 
mechanism will allow us to more easily handle similar cases in the future,  
such as support of a transactions topic.

## Rationale and alternatives

The main advantage of the approach presented here is the decoupling of the Redpanda implementation 
from Kafka internals. An alternative approach would be to migrate all the data 
from the `kafka_internal/group` topic to the new `kafka/__consumer_offsets`. 

> This, however, would bound the Redpanda implementation to the format of the data that 

Reviewer 2: Migrating `kafka_internal/group` to `kafka/__consumer_offsets` still looks very tempting. The scope of this RFC is huge: `MetadataResponse`, `ListOffsets`, `DescribeConfigs`, `DescribeLogDirs`, `AlterConfig`, `IncrementalAlterConfig`, `CreateTopic`, `Fetch`, `Produce`.

While if we migrate the group topic we'll achieve the similar effect without all this work. Also it isn't obvious that it bounds us to Kafka format, we already have an offset translation layer which maps raft log to kafka log, we may do the same for `kafka/__consumer_offsets` **when and if** it diverges from our internal format.

Kafka stores internally. Additionally we would need to add special handling 
every time Redpanda implementation would change. 

## Unresolved questions

> - how to handle CreatePartitionsAPI for `__consumer_offsets` topic ?
>   changing number of partitions in group topic would change 
>   `group_id` to `partition_id` mapping

Reviewer: ```suggestion
- How do we handle the `CreatePartitionsAPI` for the `__consumer_offsets` topic?
  Changing the number of partitions in the group topic would change the 
  `group_id` to `partition_id` mapping.
```
